src/advanced_unsafety/uninitialized.md

```diff
@@ -0,0 +1,128 @@
+# Uninitialized memory
+
+> _"I'm Nobody! Who are you? Are you — Nobody — too?"_
+>
+> — _Emily Dickinson_
+
+While we have covered [invalid values], there's another thing that is a kind of invalid value, but has nothing to do with actual bit patterns: Uninitialized memory.
+
+An easy way to think about uninitialized memory is that there's an additional value (often called `undef` using LLVM's term for it) that does not map to any concrete bit pattern, but can be introduced in the abstract machine in various ways, and makes _most_ values invalid.
```

Comment:
i really don't like connecting to LLVM's names for stuff here bc LLVM `undef` is the wrong mental model e.g. LLVM guarantees the following:

```ll
%undef = i32 undef
%undef_hi = and i32 %undef, i32 0x0000ffff
%zero = and i32 %undef_hi, i32 0xffff0000
%is_zero = icmp i32 %zero, i32 0
call void @llvm.assume(i1 %is_zero)
```

i.e. `undef` is tracked at the bit level, so the bitpattern of any non-poison value is a ternary vector.

Rust models uninitialized memory much closer to LLVM's `poison`, which behaves holistically, more like a NaN: any operation involving `poison` is poison. Thus, `undef - undef == 0` but `poison - poison == poison` (insofar that == means anything here, because `icmp poison, poison` is also poison).

(Hilariously, langref does not specify whether or not `call void @llvm.assume(i1 poison)` is UB. lmao)

Reply:
something else to discuss is that Rust does _not_ model uninit the way it LLVM models `poison`, because from LLVM's perspective operations with `poison` are well-defined *except* when: `load`, `store`, or `call` from a `ptr poison`, `br` on a `i1 poison`, or dividing by `poison`.

this is crucially different from how C++ models uninit, because C++ models almost exactly like `poison`, e.g. the following is fine in C++ and extremely not allowed in Rust:

```c++
struct X { int x; }

X x, y{0};
X* p = &x;
X* q = &y;

*q = *p;  // Copy from lvalue to uninit via X::X(const X&), y.x is now uninit.
y.x = 0;  // y is initialized again.
int z = p->x;  // Equivalent to int z;
```

Reply:
Yeah I think I'll kill this bit from the text.

Informally, people do say u BB def a l poo t, but we don't need to perpetuate that